crates/api/src/agent.rs

```diff
@@ -0,0 +1,362 @@
+    ) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        serializer.serialize_str(&t.as_micros().to_string())
+    }
+
+    pub fn deserialize<'de, D>(
+        deserializer: D,
+    ) -> Result<crate::Timestamp, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        let s: &'de str = serde::Deserialize::deserialize(deserializer)?;
+        let i: i64 = s.parse().map_err(serde::de::Error::custom)?;
+        Ok(crate::Timestamp::from_micros(i))
+    }
+}
+
+/// AgentInfo stores metadata related to agents.
+#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct AgentInfo {
+    /// The agent id.
+    pub agent: AgentId,
+
+    /// The space id.
+    pub space: SpaceId,
+
+    /// When this metadata was created.
+    #[serde(with = "serde_string_timestamp")]
+    pub created_at: Timestamp,
+
+    /// When this metadata will expire.
+    #[serde(with = "serde_string_timestamp")]
+    pub expires_at: Timestamp,
+
+    /// If `true`, this metadata is a tombstone, indicating
+    /// the agent has gone offline, and is no longer reachable.
+    pub is_tombstone: bool,
+
+    /// If set, this indicates the primary url at which this agent may
+    /// be reached. This should largely only be UNSET if this is a tombstone.
+    pub url: Option<String>,
+
+    /// The arc over which this agent claims authority.
+    pub storage_arc: BasicArc,
+}
+
+/// Signed agent information.
+#[derive(Debug)]
+pub struct AgentInfoSigned {
+    /// The decoded information associated with this agent.
+    agent_info: AgentInfo,
+
+    /// The encoded information that was signed.
+    encoded: String,
+
+    /// The signature.
+    signature: bytes::Bytes,
+}
+
+impl AgentInfoSigned {
+    /// Generate a signed agent info by signing an agent info.
+    pub async fn sign<S: Signer>(
+        signer: &S,
+        agent_info: AgentInfo,
+    ) -> std::io::Result<std::sync::Arc<Self>> {
+        let encoded = serde_json::to_string(&agent_info)?;
+        let signature = signer.sign(&agent_info, encoded.as_bytes()).await?;
+        Ok(std::sync::Arc::new(Self {
+            agent_info,
+            encoded,
+            signature,
+        }))
+    }
+
+    /// Decode a canonical json encoding of a signed agent info.
+    pub fn decode<V: Verifier>(
+        verifier: &V,
+        encoded: &[u8],
+    ) -> std::io::Result<Self> {
+        #[derive(serde::Deserialize)]
+        #[serde(rename_all = "camelCase")]
+        struct Ref {
+            agent_info: String,
+            #[serde(with = "crate::serde_bytes_base64")]
+            signature: bytes::Bytes,
+        }
+        let v: Ref = serde_json::from_slice(encoded)?;
+        let agent_info: AgentInfo = serde_json::from_str(&v.agent_info)?;
+        if !verifier.verify(&agent_info, v.agent_info.as_bytes(), &v.signature)
+        {
+            return Err(std::io::Error::other("InvalidSignature"));
```

Comment:
Nope.

- Error type enums are a huge headache without much practical benefit. I assert that they are often a pre-optimization... something we feel like we *should* do because someone might want to switch on it someday, but no one ever does, and it just makes more work in the mean time.
- kitsune is largely a networking library, so networking-type errors will need to be bubbled up to the surface. That makes std::io::Error the perfect error type for this usage. If we ever *do* find that we need to switch on a specify error type, and cannot find an equivalent std::io::ErrorKind entry for it, we can create a custom error to use with `std:io::Error::other` and people implementing our API can use the `std::io::Error::downcast` to check the specific type.